src/core/iter.rs

```diff
@@ -0,0 +1,98 @@
+// Rust Simplicity Library
+// Written in 2022 by
+//   Christian Lewe <clewe@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Iterators for DAG-like structures
+
+use std::iter::Rev;
+use std::vec::IntoIter;
+
+/// A DAG-like structure with a root,
+/// left children and right children per node,
+/// without directed cycles.
+///
+/// We use this trait to implement general DAG operations, like node traversal.
+pub trait DagLike: Sized {
+    /// Node in the DAG
+    type Node: Copy;
+    /// Additional data for internal computations
+    type Extra;
+
+    /// Return the DAG root, if the DAG is nonempty.
+    fn root(&self) -> Option<Self::Node>;
+
+    /// Return the left child of the given `node`.
+    fn left_of(&self, node: Self::Node) -> Option<Self::Node>;
```

Comment:
In 925b8cbdb0be693e1fa8d93005eb87d2a9c54846:

Would prefer this be called `left_child` and `right_child`

Reply:
Sure. I had to chose different names for these methods, because they clashed with `TermDag::get_left()` and `TermDag::get_right()`. I feel that the latter are more important and hence should have the "better" names.

Reply:
I don't think it really matters if these names collide, we only call these ones from within this trait's iterator logic, and things will be unambiguous there.

But having said this, why do we have `TermDag::get_left` etc when we already have these methods?

Reply:
The `get_left()` and `get_right()` methods are defined on the `Term`(`Dag`) itself and are there to make pattern matching less verbose. Possibly, one could use these methods for the inductive construction of `TermDag`s, too.

The `left_of()` and `right_of()` methods are defined on `DagIterable` and have a different signature: They take the dag as `self` and then the `node`. These methods are for use inside `DagIterable`. For `Term`, this distinction is important because its children depend on the program.

Therefore, I think that the two sets of methods fulfil different use cases and are both needed.

Reply:
Yeah, I see, the signatures don't match and can't be unified.